File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/4_pertahanan/1_komando_pertahanan/modals/1_misi_serangan/halaman_utama_pertempuran/pertempuran/logic/mapTexture/jalan/python/road_generator.py
#!/usr/bin/env python3
"""
Road Generator - Python Implementation
Membuat tekstur jalan berwarna hitam dengan garis putus-putus putih
"""

import sys
import json
import numpy as np
from typing import Dict, Any


class RoadGenerator:
    def __init__(self, config: Dict[str, Any]):
        self.width = config.get('width', 512)
        self.height = config.get('height', 512)
        self.road_width = config.get('roadWidth', 50)
        self.dash_length = config.get('dashLength', 20)
        self.dash_gap = config.get('dashGap', 10)
        self.road_color = self._parse_color(config.get('roadColor', '#000000'))
        self.dash_color = self._parse_color(config.get('dashColor', '#FFFFFF'))

    def _parse_color(self, color_str: str) -> tuple:
        """Parse hex color string to RGB tuple"""
        color_str = color_str.lstrip('#')
        return tuple(int(color_str[i:i+2], 16) for i in (0, 2, 4))

    def generate(self) -> np.ndarray:
        """Generate road texture"""
        # Buat canvas hitam (jalan)
        image = np.zeros((self.height, self.width, 3), dtype=np.uint8)
        image[:, :] = self.road_color

        # Hitung posisi tengah untuk garis putus-putus
        center_x = self.width // 2
        dash_width = 4  # Lebar garis putus-putus

        # Gambar garis putus-putus vertikal di tengah
        y_pos = 0
        while y_pos < self.height:
            # Gambar dash
            y_end = min(y_pos + self.dash_length, self.height)
            x_start = center_x - dash_width // 2
            x_end = center_x + dash_width // 2
            
            image[y_pos:y_end, x_start:x_end] = self.dash_color
            
            # Pindah ke posisi dash berikutnya
            y_pos += self.dash_length + self.dash_gap

        return image

    def to_json_output(self, image: np.ndarray) -> str:
        """Convert image to JSON output format"""
        result = {
            'imageData': image.flatten().tolist(),
            'width': self.width,
            'height': self.height
        }
        return json.dumps(result)


def main():
    if len(sys.argv) < 2:
        print(json.dumps({'error': 'No configuration provided'}))
        sys.exit(1)

    try:
        config = json.loads(sys.argv[1])
        generator = RoadGenerator(config)
        image = generator.generate()
        output = generator.to_json_output(image)
        print(output)
    except Exception as e:
        print(json.dumps({'error': str(e)}))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/4_pertahanan/1_komando_pertahanan/modals/1_misi_serangan/halaman_utama_pertempuran/pertempuran/logic/mapTexture/jalan/cpp/road_generator.cpp
/**
 * Road Generator - C++ Implementation
 * Membuat tekstur jalan berwarna hitam dengan garis putus-putus putih
 */

#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <cstdint>
#include <algorithm>

// Simple JSON parser untuk config
struct Color {
    uint8_t r, g, b;
};

struct RoadConfig {
    int width = 512;
    int height = 512;
    int roadWidth = 50;
    int dashLength = 20;
    int dashGap = 10;
    Color roadColor = {0, 0, 0};
    Color dashColor = {255, 255, 255};
};

class RoadGenerator {
private:
    RoadConfig config;
    std::vector<uint8_t> imageData;

    Color parseColor(const std::string& hexColor) {
        Color color;
        std::string hex = hexColor;
        if (hex[0] == '#') hex = hex.substr(1);
        
        color.r = std::stoi(hex.substr(0, 2), nullptr, 16);
        color.g = std::stoi(hex.substr(2, 2), nullptr, 16);
        color.b = std::stoi(hex.substr(4, 2), nullptr, 16);
        
        return color;
    }

    void setPixel(int x, int y, const Color& color) {
        if (x >= 0 && x < config.width && y >= 0 && y < config.height) {
            int index = (y * config.width + x) * 3;
            imageData[index] = color.r;
            imageData[index + 1] = color.g;
            imageData[index + 2] = color.b;
        }
    }

public:
    RoadGenerator(const RoadConfig& cfg) : config(cfg) {
        imageData.resize(config.width * config.height * 3);
    }

    void generate() {
        // Fill dengan warna jalan (hitam)
        for (int y = 0; y < config.height; y++) {
            for (int x = 0; x < config.width; x++) {
                setPixel(x, y, config.roadColor);
            }
        }

        // Gambar garis putus-putus di tengah
        int centerX = config.width / 2;
        int dashWidth = 4;
        int yPos = 0;

        while (yPos < config.height) {
            int yEnd = std::min(yPos + config.dashLength, config.height);
            int xStart = centerX - dashWidth / 2;
            int xEnd = centerX + dashWidth / 2;

            // Gambar dash
            for (int y = yPos; y < yEnd; y++) {
                for (int x = xStart; x < xEnd; x++) {
                    setPixel(x, y, config.dashColor);
                }
            }

            yPos += config.dashLength + config.dashGap;
        }
    }

    std::string toJSON() const {
        std::ostringstream oss;
        oss << "{\"imageData\":[";
        
        for (size_t i = 0; i < imageData.size(); i++) {
            if (i > 0) oss << ",";
            oss << static_cast<int>(imageData[i]);
        }
        
        oss << "],\"width\":" << config.width 
            << ",\"height\":" << config.height << "}";
        
        return oss.str();
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "{\"error\":\"No configuration provided\"}" << std::endl;
        return 1;
    }

    try {
        // Parse config dari JSON (simplified)
        RoadConfig config;
        // TODO: Implement proper JSON parsing
        // Untuk sekarang gunakan default values
        
        RoadGenerator generator(config);
        generator.generate();
        
        std::cout << generator.toJSON() << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "{\"error\":\"" << e.what() << "\"}" << std::endl;
        return 1;
    }

    return 0;
}
